primitives/state-machine/src/trie_backend_essence.rs

```diff
@@ -58,12 +59,12 @@ pub trait Storage<H: Hasher>: Send + Sync {
 
 /// Local cache for child root.
 #[cfg(feature = "std")]
-pub(crate) struct Cache {
-	pub child_root: HashMap<Vec<u8>, Option<Vec<u8>>>,
+pub(crate) struct Cache<H> {
+	pub child_root: HashMap<Vec<u8>, Option<H>>,
```

Comment:
The Switch from encoded root to decoded is fine as long as we only have one kind of root type.
Just mentioning, I recently looked a bit again at how to have multiple ct kind, and thought state machine should abstract from it's H trait.
In fact even if removing H::Out usage, H trait would still be required to expose some host function (a hasher for the top and default child trie at least).
But I am fine with the change.
Actually looks better this way.

Reply:
I mean nothing is currently prepared for this and we have enough open construction sites that we don't need to consider this yet. This also should lead to one less copy each time we access the hash :)